### PR TITLE
#169 Fix the issue with highlight was not displayed correct in some s…

### PR DESCRIPTION
**Pull request by fharms**
- Fix the bug with highlight was not displayed correct in some situation.
- Fix the issue with isCameSimpleExpressionUsedAsPredicate didn't work correctly

We need to add a lot more test cases. It's to easy to break the functionality because the IDEA Annotator and Text range  behave different depended on which file type and type of expression  simple / Predicate

### CODE REVIEW

File: camel-idea-plugin/src/main/java/org/apache/camel/idea/annotator/CamelSimpleAnnotator.java [hunk 1]
```diff
@@ -69,9 +69,14 @@ void validateText(@NotNull PsiElement element, @NotNull AnnotationHolder holder,
                     if (!result.isSuccess()) {
                         String error = result.getShortError();
                         TextRange range = element.getTextRange();
+                        int startIdx = result.getIndex() == 0 ? text.indexOf("$") : result.getIndex();
```
Comment (davsclaus) [inline camel-idea-plugin/src/main/java/org/apache/camel/idea/annotator/CamelSimpleAnnotator.java]: Can you add some code comments to this code, its hard to follow what goes on, with the various things. Yes writing code with indexOf and substrings tend to be a bit unclear.

File: camel-idea-plugin/src/main/java/org/apache/camel/idea/annotator/CamelSimpleAnnotator.java [hunk 2]
```diff
@@ -69,9 +69,14 @@ void validateText(@NotNull PsiElement element, @NotNull AnnotationHolder holder,
...
+
+                        int endIdx = text.indexOf("}", startIdx);
+                        endIdx = endIdx == -1 ? text.indexOf(" ", startIdx) - 1 : endIdx;
+
+                        int propertyLength = (((endIdx < 0 ? text.length() - 1 : endIdx) + 1) - startIdx) + 1;
                         if (result.getIndex() > 0) {
-                            // use -1 to skip the last quote sign
-                            range = TextRange.create(range.getStartOffset() + result.getIndex(), range.getEndOffset() - 1);
+                            range = TextRange.create(range.getStartOffset() + startIdx + 1, range.getStartOffset() + startIdx + propertyLength);
```
Comment (davsclaus) [inline camel-idea-plugin/src/main/java/org/apache/camel/idea/annotator/CamelSimpleAnnotator.java]: We had some stacktraces before when we had this kind of code, where the range would end up having start index > end index. So we should have some safe guard code that checks for this just in case. When it happened before it can happen again.

File: camel-idea-plugin/src/test/java/org/apache/camel/idea/annotator/CamelSimpleAnnotatorTestIT.java [hunk 3]
```diff
@@ -61,14 +61,68 @@ public void testAnnotatorSimpleValidation() {
...
+            + "        }\n"
+            + "    }";
+    }
+
+    private String getJavaWithPredicate() {
+        return "import org.apache.camel.builder.RouteBuilder;\n"
+            + "public class MyRouteBuilder extends RouteBuilder {\n"
+            + "        public void configure() throws Exception {\n"
+            + "              from(\"direct:start\")\n"
+            + "                .loopDoWhile(simple(\"${body.length} <error descr=\"Unknown function: xrouteId\">=!=</error> 12\"))\n"
```
Comment (davsclaus) [inline camel-idea-plugin/src/test/java/org/apache/camel/idea/annotator/CamelSimpleAnnotatorTestIT.java]: Can you add a 2nd test where you do the variation of this as

    .filter().simple("xxx")

You genererally have 2 ways in java dsl, either as parameter

    .filter ( simple("xxx") )

Or as chained method call

    .filter . simple("xxx")

And that psi parser code is different depending on one other the other are used.

File: camel-idea-plugin/src/main/java/org/apache/camel/idea/annotator/CamelSimpleAnnotator.java (hunk 2 above)
Comment (fharms) [inline camel-idea-plugin/src/main/java/org/apache/camel/idea/annotator/CamelSimpleAnnotator.java]: We can implement more safeguards and more comment. But the previous solution had the limit that it always assumed it was to the end of the element and to much was highlighted

With the new solution I try to solve 2 issue : 
1. Find the first " " or } after the getIndex
2. if the expression is never ended with " " or "}" we should use the range.getEndOffset()

I will try to construct in another way so it's more readable

File: camel-idea-plugin/src/test/java/org/apache/camel/idea/annotator/CamelSimpleAnnotatorTestIT.java (hunk 3 above)
Comment (fharms) [inline camel-idea-plugin/src/test/java/org/apache/camel/idea/annotator/CamelSimpleAnnotatorTestIT.java]: Will do.

**Comment by fharms** [comment]
Close this because there is still on going work

**Comment by fharms** [comment]
@davsclaus let's take another round with this :-)

File: camel-idea-plugin/src/main/java/org/apache/camel/idea/util/CamelIdeaUtils.java [hunk 4]
```diff
@@ -168,7 +168,7 @@ public static boolean isCameSimpleExpressionUsedAsPredicate(PsiElement element)
                         element = element.getParent();
                     }
                     if (element instanceof PsiMethodCallExpression) {
-                        exp = (PsiMethodCallExpression) element;
+                        exp = PsiTreeUtil.getParentOfType(element.getParent(), PsiMethodCallExpression.class);
```
Comment (davsclaus) [inline camel-idea-plugin/src/main/java/org/apache/camel/idea/util/CamelIdeaUtils.java]: this looks a bit wrong, before it was a type cast, you you get the first parent or grand parents that is a method call.

File: camel-idea-plugin/src/main/java/org/apache/camel/idea/annotator/CamelSimpleAnnotator.java [hunk 5]
```diff
@@ -49,7 +50,7 @@ boolean isEnabled() {
      */
     void validateText(@NotNull PsiElement element, @NotNull AnnotationHolder holder, @NotNull String text) {
         boolean hasSimple = text.contains("${") || text.contains("$simple{");
-        if (hasSimple && CamelIdeaUtils.isCamelSimpleExpression(element)) {
+        if (hasSimple || CamelIdeaUtils.isCamelSimpleExpression(element)) {
```
Comment (davsclaus) [inline camel-idea-plugin/src/main/java/org/apache/camel/idea/annotator/CamelSimpleAnnotator.java]: This should in fact just only be `isCamelSimpleExpression` as that checks if its simple or not. The has simple was to test for simple functions, but you can do simple without functions, such as `5 < 10'. But not so sure if they make so much sense.

**Comment by davsclaus** [comment]
@fharms there is a merge conflict now due the travis file has been merged

File: camel-idea-plugin/src/main/java/org/apache/camel/idea/util/CamelIdeaUtils.java (hunk 4 above)
Comment (fharms) [inline camel-idea-plugin/src/main/java/org/apache/camel/idea/util/CamelIdeaUtils.java]: previous it didn't go far enough back in the PSI tree to find parent method and therefore it didn't see the simple method was an input predicate to another method.

File: camel-idea-plugin/src/main/java/org/apache/camel/idea/annotator/CamelSimpleAnnotator.java (hunk 5 above)
Comment (fharms) [inline camel-idea-plugin/src/main/java/org/apache/camel/idea/annotator/CamelSimpleAnnotator.java]: I previously removed this and you made comment, that it would not work for expression like this
`.log("Result from query ${body}")`

Is the CamelIdeaUtils.isCamelSimpleExpression(element) handling this case also?

Comment (fharms) [inline camel-idea-plugin/src/main/java/org/apache/camel/idea/annotator/CamelSimpleAnnotator.java]: If I remove the 

> boolean hasSimple = text.contains("${") || text.contains("$simple{");

the test will failed for expressions like this
> .log("Result from query ${body}")

Comment (davsclaus) [inline camel-idea-plugin/src/main/java/org/apache/camel/idea/annotator/CamelSimpleAnnotator.java]: Yes but its still wrong. We need to add special support for log. Its only for simple / <simple> currently.

Otherwise you end up validating all kind of non Camel stuff as well.

My PR will correct this and have a TODO for log

**Comment by fharms** [comment]
@davsclaus see my comments. I might be wrong or my test case is wrong, but previous it didn't work with code like this:

```java
 from("direct:start")
                .loopDoWhile(simple("body.length !s1= 12"))
                .to("mock:loop")
                .transform(body().append("A"))
                .end()
                .to("mock:result");
```

**Comment by davsclaus** [comment]
Camel simple is only for `${ }` functions not without that. Its deprecated without functions and the old parser is being removed from Camel.

File: camel-idea-plugin/src/main/java/org/apache/camel/idea/annotator/CamelSimpleAnnotator.java (hunk 5 above)
Comment (fharms) [inline camel-idea-plugin/src/main/java/org/apache/camel/idea/annotator/CamelSimpleAnnotator.java]: So if do something like this in CamelIdeaUtils.isCamelSimpleExpression(element) we should be safe?

>  if (IdeaUtils.isFromJavaMethodCall(element, "simple", "log")) {
            return true;
        }

File: camel-idea-plugin/src/main/java/org/apache/camel/idea/util/CamelIdeaUtils.java (hunk 4 above)
Comment (fharms) [inline camel-idea-plugin/src/main/java/org/apache/camel/idea/util/CamelIdeaUtils.java]: Okay, revert my change here
